predicates/predicates.go

```diff
@@ -0,0 +1,118 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package predicates
+
+import (
+	"reflect"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/redhat-appstudio/release-service/api/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// ReleasePlanPairUpdatePredicate returns a predicate which returns true when a ReleasPlan
+// is created, deleted, or when the target or application of one changes.
+func ReleasePlanPairUpdatePredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(createEvent event.CreateEvent) bool {
+			return true
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+			return true
+		},
+		GenericFunc: func(genericEvent event.GenericEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return hasApplicationChanged(e.ObjectOld, e.ObjectNew) || hasTargetChanged(e.ObjectOld, e.ObjectNew)
+		},
+	}
+}
+
+// ReleasePlanAdmissionPairUpdatePredicate returns a predicate which returns true when a
+// ReleasePlanAdmission is created, deleted, or when the origin or applications of one changes.
+func ReleasePlanAdmissionPairUpdatePredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(createEvent event.CreateEvent) bool {
+			return true
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+			return true
+		},
+		GenericFunc: func(genericEvent event.GenericEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return haveApplicationsChanged(e.ObjectOld, e.ObjectNew) || hasOriginChanged(e.ObjectOld, e.ObjectNew)
+		},
+	}
+}
+
+// hasApplicationChanged returns true if the Spec.Application value changed between
+// the two releasePlans.
+func hasApplicationChanged(objectOld, objectNew client.Object) bool {
+	if releasePlanOld, ok := objectOld.(*v1alpha1.ReleasePlan); ok {
+		if releasePlanNew, ok := objectNew.(*v1alpha1.ReleasePlan); ok {
+			return releasePlanOld.Spec.Application != releasePlanNew.Spec.Application
+		}
+	}
+
+	return false
+}
+
+// haveApplicationsChanged returns true if the Spec.Applications value is different
+// between the two releasePlanAdmissions.
+func haveApplicationsChanged(objectOld, objectNew client.Object) bool {
+	if releasePlanAdmissionOld, ok := objectOld.(*v1alpha1.ReleasePlanAdmission); ok {
+		if releasePlanAdmissionNew, ok := objectNew.(*v1alpha1.ReleasePlanAdmission); ok {
+			oldApps := releasePlanAdmissionOld.Spec.Applications
+			newApps := releasePlanAdmissionNew.Spec.Applications
+			slices.Sort(oldApps)
+			slices.Sort(newApps)
```

Comment:
Slices are already ordered. I would just compare directly probably. Have you tried?

Reply:
I removed the sort. Let's leave this comment open though until I test it